auto_submit/lib/requests/check_pull_request.dart

```diff
@@ -27,23 +27,56 @@ class CheckPullRequest extends RequestHandler {
     final String rawBody = await request.readAsString();
     final body = json.decode(rawBody) as Map<String, dynamic>;
     final PullRequest pullRequest = PullRequest.fromJson(body['pull_request']);
+    final RepositorySlug slug = pullRequest.base!.repo!.slug();
 
     final GithubService gitHub = await config.createGithubService();
     final _AutoMergeQueryResult queryResult = await _parseQueryData(pullRequest, gitHub);
-    if (await shouldMergePullRequest(queryResult)) {
+    if (await shouldMergePullRequest(queryResult, slug, gitHub)) {
       // TODO(Kristin): Keep pulling pubsub queue, https://github.com/flutter/flutter/issues/98704
 
-    } else {
+    } else if (queryResult.shouldRemoveLabel) {
+      log.info('Removing label for commit: ${queryResult.sha}');
+      await _removeLabel();
       return Response.ok(jsonEncode(<String, String>{}));
     }
 
     return Response.ok(rawBody);
   }
 
   /// Check if the pull request should be merged.
-  Future<bool> shouldMergePullRequest(_AutoMergeQueryResult queryResult) async {
-    // TODO(Kristin): Implement shouldMergePullRequest, https://github.com/flutter/flutter/issues/98707
+  ///
+  /// A pull request should be merged on either cases:
+  /// 1) All tests have finished running and satified basic merge requests
+  /// 2) Not all tests finish but this is a clean revert of the Tip of Tree (TOT) commit.
+  Future<bool> shouldMergePullRequest(
+      _AutoMergeQueryResult queryResult, RepositorySlug slug, GithubService github) async {
+    if (queryResult.shouldMerge) {
+      return true;
+    }
+    // If the PR is a revert of the tot commit, merge without waiting for checks passing.
+    return await isTOTRevert(queryResult.sha, slug, github);
+  }
+
+  /// Check if the `commitSha` is a clean revert of TOT commit.
+  ///
+  /// By comparing the current commit with second TOT commit, an empty `files` in
```

Comment:
It's still unclear to me why we need the second commit, and not the tot

Reply:
Hi Casey, I will give you an example here. 
Let's assume we have 3 commits. Commit1 is the TOT, commit2 is the second commit, and commit0 is the commit to be check here and is also the revert of commit1(TOT).
Let's then assume commit2 made a change to add file A,  while commit1 made a change to add file B. Since commit0 is the top, it will include all changes made in previous commits. Since commit0 is the revert commit of TOT, it will revert the changes made by commit1, that is, revert adding file B. As a result, commit0 now only contains the previous changes of adding file A which is made by commit2.
Then, when we compare commit0 to commit1, since commit1 added file B but commit0 didn't, the changes shown in the file will be in the opposite way. When we compare commit0 and commit2, since they both added file A and commit0 didn't contain any other changes made by commit1 now, the file will be empty and we can get to know that commit0 is a revert of TOT.

Reply:
Can you put this in a chart? To recap, you're saying we have:

Tree: 2 -> 1
PR has 0

If the PR is a ToT revert, then when merged we would have
Tree: 2 -> 1 -> 0
In essence, the tree should be 2 equivalent to what is with 2

PS: I'm not sure i'm capturing the content correctly, but just to give you and idea on how to clearly communicate this

Reply:
Hi Casey, thank you fro giving me an example here. I think you got my idea, but I will also try to use chart to explain it more clearly:

Tree: 2 -> 1
Changes: 2( + fileA), 1(+ fileB)
PR has 0

If the PR 0 is a clean ToT revert:
changes for 0 is: 0(revert +fileB)

then when merged we would have
Tree: 2 -> 1 -> 0
In essence, the status of 0 is equivalent to what is with 2, but different with 1